File: gawk-server/wire/parity_test.go
package wire

import (
	"bytes"
	"encoding/hex"
	"errors"
	"math/rand"
	"testing"
)

// --- GF(2^8) arithmetic ----------------------------------------------------

func TestGFMulIdentitiesAndInverse(t *testing.T) {
	for a := 0; a < 256; a++ {
		if got := gfMul(uint8(a), 0); got != 0 {
			t.Fatalf("gfMul(%d, 0) = %d, want 0", a, got)
		}
		if got := gfMul(uint8(a), 1); got != uint8(a) {
			t.Fatalf("gfMul(%d, 1) = %d, want %d", a, got, a)
		}
	}
	// Every non-zero element has a multiplicative inverse.
	for a := 1; a < 256; a++ {
		inv := gfDiv(1, uint8(a))
		if got := gfMul(uint8(a), inv); got != 1 {
			t.Fatalf("a=%d * inv=%d = %d, want 1", a, inv, got)
		}
	}
}

func TestGFPowIsDistinctBelow255(t *testing.T) {
	// The Q coefficients are g^i. They must be distinct over the supported
	// data-chunk range or the 2-erasure solve divides by zero. This is
	// exactly why MaxParityDataChunks is 255.
	seen := make(map[uint8]int, 255)
	for i := 0; i < MaxParityDataChunks; i++ {
		c := gfPow2(i)
		if prev, dup := seen[c]; dup {
			t.Fatalf("g^%d == g^%d == %d: coefficients repeat inside the supported range", i, prev, c)
		}
		seen[c] = i
	}
	// And it genuinely wraps right after, which is the guard's justification.
	if gfPow2(0) != gfPow2(255) {
		t.Fatal("expected g^0 == g^255 (period 255) — the guard rationale is wrong")
	}
}

// --- Parity computation ----------------------------------------------------

func TestComputeParityShape(t *testing.T) {
	chunks := [][]byte{{1, 2, 3, 4}, {5, 6, 7, 8}, {9, 10}}
	p, err := ComputeParity(chunks, 2)
	if err != nil {
		t.Fatalf("ComputeParity: %v", err)
	}
	if len(p) != 2 {
		t.Fatalf("got %d symbols, want 2", len(p))
	}
	// Symbols are as long as the LONGEST chunk (short chunks zero-pad), which
	// is what lets the viewer derive the chunk length from a parity payload.
	for i, s := range p {
		if len(s) != 4 {
			t.Fatalf("symbol %d has length %d, want 4", i, len(s))
		}
	}
	// P is the plain XOR — the prefix property the whole per-subscriber
	// design rests on (docs/34 §4.1).
	wantP := []byte{1 ^ 5 ^ 9, 2 ^ 6 ^ 10, 3 ^ 7, 4 ^ 8}
	if !bytes.Equal(p[0], wantP) {
		t.Fatalf("P = %x, want %x", p[0], wantP)
	}
}

func TestComputeParityPrefixProperty(t *testing.T) {
	chunks := [][]byte{{0xde, 0xad}, {0xbe, 0xef}, {0x12, 0x34}}
	one, err := ComputeParity(chunks, 1)
	if err != nil {
		t.Fatalf("ComputeParity(k=1): %v", err)
	}
	two, err := ComputeParity(chunks, 2)
	if err != nil {
		t.Fatalf("ComputeParity(k=2): %v", err)
	}
	if !bytes.Equal(one[0], two[0]) {
		t.Fatalf("k=1 symbol %x != k=2 symbol 0 %x: prefix property broken", one[0], two[0])
	}
}

func TestComputeParitySingleChunkEmitsOneSymbol(t *testing.T) {
	// n=1: P duplicates the chunk (genuine protection against its loss); a
	// second symbol would add nothing, so min(k, n) applies.
	chunks := [][]byte{{7, 7, 7}}
	p, err := ComputeParity(chunks, 2)
	if err != nil {
		t.Fatalf("ComputeParity: %v", err)
	}
	if len(p) != 1 {
		t.Fatalf("got %d symbols for n=1, want 1", len(p))
	}
	if !bytes.Equal(p[0], chunks[0]) {
		t.Fatalf("P = %x, want the chunk itself %x", p[0], chunks[0])
	}
}

func TestComputeParityRejectsTooManyChunks(t *testing.T) {
	chunks := make([][]byte, MaxParityDataChunks+1)
	for i := range chunks {
		chunks[i] = []byte{1}
	}
	if _, err := ComputeParity(chunks, 2); !errors.Is(err, ErrParityUnsupported) {
		t.Fatalf("err = %v, want ErrParityUnsupported for n > %d", err, MaxParityDataChunks)
	}
}

func TestComputeParityRejectsBadK(t *testing.T) {
	chunks := [][]byte{{1}, {2}}
	for _, k := range []int{-1, MaxParitySymbols + 1} {
		if _, err := ComputeParity(chunks, k); err == nil {
			t.Fatalf("k=%d accepted, want error", k)
		}
	}
	if p, err := ComputeParity(chunks, 0); err != nil || p != nil {
		t.Fatalf("k=0 should be a clean no-op, got %v / %v", p, err)
	}
}

// --- Recovery: the exhaustive property test --------------------------------

// TestRecoverAllErasurePairs is the load-bearing correctness proof: for every
// supported n and EVERY pair of erasure positions among the n+2 transmitted
// chunks, recovery must reproduce the original bytes exactly.
func TestRecoverAllErasurePairs(t *testing.T) {
	rnd := rand.New(rand.NewSource(1))
	for _, n := range []int{1, 2, 3, 9, 17, 64, 254, 255} {
		orig, frameBytes := makeFrame(rnd, n, 32)
		k := 2
		if n < 2 {
			k = 1
		}
		parity, err := ComputeParity(orig, k)
		if err != nil {
			t.Fatalf("n=%d: ComputeParity: %v", n, err)
		}
		total := n + len(parity)
		for a := 0; a < total; a++ {
			for b := a; b < total; b++ {
				// a == b models a single erasure.
				chunks, par := eraseAt(orig, parity, a, b)
				lost := 0
				for _, c := range chunks {
					if c == nil {
						lost++
					}
				}
				// Recoverability is data erasures vs. SURVIVING parity, not
				// vs. the parity originally sent — erasing a symbol removes
				// exactly as much repair capacity as erasing a data chunk
				// consumes.
				survivingParity := 0
				for _, p := range par {
					if p != nil {
						survivingParity++
					}
				}
				err := RecoverChunks(chunks, par, frameBytes)
				if lost > survivingParity {
					if err == nil {
						t.Fatalf("n=%d erasures (%d,%d): recovered %d data losses with %d surviving symbols", n, a, b, lost, survivingParity)
					}
					continue
				}
				if err != nil {
					t.Fatalf("n=%d erasures (%d,%d): RecoverChunks: %v", n, a, b, err)
				}
				for i := range orig {
					if !bytes.Equal(chunks[i], orig[i]) {
						t.Fatalf("n=%d erasures (%d,%d): chunk %d = %x, want %x", n, a, b, i, chunks[i], orig[i])
					}
				}
			}
		}
	}
}

func TestRecoverPreservesShortFinalChunk(t *testing.T) {
	// The last chunk is shorter than the rest. frameBytes is the only thing
	// that says by how much, so losing exactly that chunk is the case the
	// header field exists for.
	orig := [][]byte{{1, 2, 3, 4}, {5, 6, 7, 8}, {9}}
	frameBytes := 9
	parity, err := ComputeParity(orig, 2)
	if err != nil {
		t.Fatalf("ComputeParity: %v", err)
	}
	chunks := [][]byte{orig[0], orig[1], nil}
	if err := RecoverChunks(chunks, parity, frameBytes); err != nil {
		t.Fatalf("RecoverChunks: %v", err)
	}
	if !bytes.Equal(chunks[2], []byte{9}) {
		t.Fatalf("recovered final chunk = %x, want 09 (padding must be trimmed)", chunks[2])
	}
}

func TestRecoverFailsCleanlyWithoutParity(t *testing.T) {
	orig := [][]byte{{1, 2}, {3, 4}}
	chunks := [][]byte{nil, orig[1]}
	if err := RecoverChunks(chunks, nil, 4); err == nil {
		t.Fatal("recovery without any parity should fail")
	}
	// A frame with nothing missing needs no parity and must not error.
	full := [][]byte{orig[0], orig[1]}
	if err := RecoverChunks(full, nil, 4); err != nil {
		t.Fatalf("complete frame: %v", err)
	}
}

func TestRecoverRejectsInconsistentFrameBytes(t *testing.T) {
	orig := [][]byte{{1, 2, 3, 4}, {5, 6}}
	parity, _ := ComputeParity(orig, 2)
	chunks := [][]byte{nil, orig[1]}
	// frameBytes far larger than n * chunkLen is structurally impossible.
	if err := RecoverChunks(chunks, parity, 9999); err == nil {
		t.Fatal("inconsistent frameBytes accepted, want error")
	}
}

// --- Wire encoding ---------------------------------------------------------

func TestParityChunkRoundTrip(t *testing.T) {
	h := ParityChunkHeader{FrameID: 0x01020304, ParityIndex: 1, ChunkCount: 9, FrameBytes: 8640}
	payload := bytes.Repeat([]byte{0xa5}, 64)
	dgram, err := AppendParityChunk(nil, h, payload)
	if err != nil {
		t.Fatalf("AppendParityChunk: %v", err)
	}
	if len(dgram) != ParityChunkHeaderSize+len(payload) {
		t.Fatalf("len = %d, want %d", len(dgram), ParityChunkHeaderSize+len(payload))
	}
	got, gotPayload, err := ParseParityChunk(dgram)
	if err != nil {
		t.Fatalf("ParseParityChunk: %v", err)
	}
	if got != h {
		t.Fatalf("header = %+v, want %+v", got, h)
	}
	if !bytes.Equal(gotPayload, payload) {
		t.Fatalf("payload = %x, want %x", gotPayload, payload)
	}
}

// TestParityChunkGoldenVector pins the exact bytes. The TS and gawk-broadcast
// mirrors assert the same hex.
func TestParityChunkGoldenVector(t *testing.T) {
	h := ParityChunkHeader{FrameID: 0x01020304, ParityIndex: 1, ChunkCount: 9, FrameBytes: 8640}
	dgram, err := AppendParityChunk(nil, h, []byte{0xde, 0xad, 0xbe, 0xef})
	if err != nil {
		t.Fatalf("AppendParityChunk: %v", err)
	}
	const want = "010e01020304010009000021c0deadbeef"
	if got := hex.EncodeToString(dgram); got != want {
		t.Fatalf("golden vector mismatch:\n got %s\nwant %s", got, want)
	}
}

// TestParityChunkFullPayloadBoundary pins the common case, not an edge case:
// a parity symbol over full-size data chunks is MaxChunkPayload bytes, and
// the resulting datagram must fit MaxDatagramSize. This is why the header is
// 13 bytes and not 20 (docs/34 §4.2).
func TestParityChunkFullPayloadBoundary(t *testing.T) {
	payload := bytes.Repeat([]byte{0x5a}, MaxChunkPayload)
	dgram, err := AppendParityChunk(nil, ParityChunkHeader{FrameID: 1, ChunkCount: 9, FrameBytes: 9 * MaxChunkPayload}, payload)
	if err != nil {
		t.Fatalf("AppendParityChunk: %v", err)
	}
	if len(dgram) > MaxDatagramSize {
		t.Fatalf("full-payload parity datagram is %d bytes, exceeds MaxDatagramSize %d", len(dgram), MaxDatagramSize)
	}
	if len(dgram) != ParityChunkHeaderSize+MaxChunkPayload {
		t.Fatalf("len = %d, want %d", len(dgram), ParityChunkHeaderSize+MaxChunkPayload)
	}
	if _, p, err := ParseParityChunk(dgram); err != nil || len(p) != MaxChunkPayload {
		t.Fatalf("round trip at boundary failed: %v, payload %d", err, len(p))
	}
}

func TestParityChunkRejectsMalformed(t *testing.T) {
	good, err := AppendParityChunk(nil, ParityChunkHeader{FrameID: 1, ChunkCount: 4, FrameBytes: 16}, []byte{1, 2, 3, 4})
	if err != nil {
		t.Fatalf("AppendParityChunk: %v", err)
	}
	for name, mutate := range map[string]func([]byte) []byte{
		"short":        func(b []byte) []byte { return b[:ParityChunkHeaderSize-1] },
		"bad version":  func(b []byte) []byte { c := clone(b); c[0] = 0x02; return c },
		"bad type":     func(b []byte) []byte { c := clone(b); c[1] = TypeVideoChunk; return c },
		"zero count":   func(b []byte) []byte { c := clone(b); c[7], c[8] = 0, 0; return c },
		"count > max":  func(b []byte) []byte { c := clone(b); c[7], c[8] = 0xff, 0xff; return c },
		"bad index":    func(b []byte) []byte { c := clone(b); c[6] = MaxParitySymbols; return c },
		"oversize pay": func(b []byte) []byte { return append(clone(b), make([]byte, MaxDatagramSize)...) },
	} {
		if _, _, err := ParseParityChunk(mutate(good)); err == nil {
			t.Fatalf("%s: accepted, want error", name)
		}
	}
}

func TestAppendParityChunkRejectsOversizePayload(t *testing.T) {
	_, err := AppendParityChunk(nil, ParityChunkHeader{FrameID: 1, ChunkCount: 2, FrameBytes: 4}, make([]byte, MaxChunkPayload+1))
	if !errors.Is(err, ErrPayloadTooLarge) {
		t.Fatalf("err = %v, want ErrPayloadTooLarge", err)
	}
}

// --- RelayCapabilities -----------------------------------------------------

func TestRelayCapabilitiesRoundTrip(t *testing.T) {
	c := RelayCapabilities{Flags: CapParityChunks, ParityLevel: 2}
	dgram, err := AppendRelayCapabilities(nil, c)
	if err != nil {
		t.Fatalf("AppendRelayCapabilities: %v", err)
	}
	if len(dgram) != RelayCapabilitiesSize {
		t.Fatalf("len = %d, want %d", len(dgram), RelayCapabilitiesSize)
	}
	got, err := ParseRelayCapabilities(dgram)
	if err != nil {
		t.Fatalf("ParseRelayCapabilities: %v", err)
	}
	if got != c {
		t.Fatalf("got %+v, want %+v", got, c)
	}
}

func TestRelayCapabilitiesGoldenVector(t *testing.T) {
	dgram, err := AppendRelayCapabilities(nil, RelayCapabilities{Flags: CapParityChunks, ParityLevel: 2})
	if err != nil {
		t.Fatalf("AppendRelayCapabilities: %v", err)
	}
	const want = "010f000102"
	if got := hex.EncodeToString(dgram); got != want {
		t.Fatalf("golden vector mismatch:\n got %s\nwant %s", got, want)
	}
}

func TestRelayCapabilitiesRejectsMalformed(t *testing.T) {
	good, _ := AppendRelayCapabilities(nil, RelayCapabilities{Flags: CapParityChunks, ParityLevel: 1})
	for name, bad := range map[string][]byte{
		"short":       good[:RelayCapabilitiesSize-1],
		"long":        append(clone(good), 0),
		"bad version": {0x02, TypeRelayCapabilities, 0, 1, 1},
		"bad type":    {Version, TypeVideoChunk, 0, 1, 1},
		"bad level":   {Version, TypeRelayCapabilities, 0, 1, MaxParitySymbols + 1},
	} {
		if _, err := ParseRelayCapabilities(bad); err == nil {
			t.Fatalf("%s: accepted, want error", name)
		}
	}
}

func TestParityFuzzNeverPanics(t *testing.T) {
	rnd := rand.New(rand.NewSource(7))
	buf := make([]byte, 64)
	for i := 0; i < 20000; i++ {
		n := rnd.Intn(len(buf))
		rnd.Read(buf[:n])
		_, _, _ = ParseParityChunk(buf[:n])
		_, _ = ParseRelayCapabilities(buf[:n])
	}
}

// --- helpers ---------------------------------------------------------------

func clone(b []byte) []byte { return append([]byte(nil), b...) }

// makeFrame builds n chunks of chunkLen bytes with a short final chunk, and
// returns the total frame length they encode.
func makeFrame(rnd *rand.Rand, n, chunkLen int) ([][]byte, int) {
	chunks := make([][]byte, n)
	for i := range chunks {
		l := chunkLen
		if i == n-1 {
			l = chunkLen/2 + 1
		}
		chunks[i] = make([]byte, l)
		rnd.Read(chunks[i])
	}
	return chunks, (n-1)*chunkLen + len(chunks[n-1])
}

// eraseAt returns copies of chunks/parity with positions a and b removed,
// where positions [0,n) index data chunks and [n, n+k) index parity symbols.
func eraseAt(chunks, parity [][]byte, a, b int) ([][]byte, [][]byte) {
	n := len(chunks)
	c := append([][]byte(nil), chunks...)
	p := append([][]byte(nil), parity...)
	for _, pos := range []int{a, b} {
		if pos < n {
			c[pos] = nil
		} else {
			p[pos-n] = nil
		}
	}
	return c, p
}

File: gawk-app/src/transport/parity.test.ts
import { describe, expect, it } from 'vitest';
import {
  CAP_PARITY_CHUNKS,
  MAX_PARITY_DATA_CHUNKS,
  MAX_PARITY_SYMBOLS,
  PARITY_CHUNK_HEADER_SIZE,
  RELAY_CAPABILITIES_SIZE,
  computeParity,
  encodeParityChunk,
  encodeRelayCapabilities,
  gfMul,
  gfPow2,
  parseParityChunk,
  parseRelayCapabilities,
  recoverChunks,
} from './parity';
import { MAX_CHUNK_PAYLOAD, MAX_DATAGRAM_SIZE, TYPE_VIDEO_CHUNK, WIRE_VERSION, WireError } from './wire';

const hex = (b: Uint8Array) =>
  Array.from(b)
    .map((v) => v.toString(16).padStart(2, '0'))
    .join('');

describe('GF(2^8)', () => {
  it('has 0 and 1 as annihilator and identity', () => {
    for (let a = 0; a < 256; a++) {
      expect(gfMul(a, 0)).toBe(0);
      expect(gfMul(a, 1)).toBe(a);
    }
  });

  it('gives distinct Q coefficients across the supported range', () => {
    // This is exactly why MAX_PARITY_DATA_CHUNKS is 255: past it the
    // coefficients repeat and the 2-erasure solve divides by zero.
    const seen = new Map<number, number>();
    for (let i = 0; i < MAX_PARITY_DATA_CHUNKS; i++) {
      const c = gfPow2(i);
      expect(seen.has(c), `g^${i} collides with g^${seen.get(c)}`).toBe(false);
      seen.set(c, i);
    }
    expect(gfPow2(0)).toBe(gfPow2(255));
  });
});

describe('computeParity', () => {
  it('sizes symbols to the longest chunk and makes P the plain XOR', () => {
    const chunks = [Uint8Array.of(1, 2, 3, 4), Uint8Array.of(5, 6, 7, 8), Uint8Array.of(9, 10)];
    const p = computeParity(chunks, 2);
    expect(p).toHaveLength(2);
    for (const s of p) expect(s.length).toBe(4);
    expect(Array.from(p[0])).toEqual([1 ^ 5 ^ 9, 2 ^ 6 ^ 10, 3 ^ 7, 4 ^ 8]);
  });

  it('keeps the prefix property: the k=1 symbol is the k=2 symbol 0', () => {
    const chunks = [Uint8Array.of(0xde, 0xad), Uint8Array.of(0xbe, 0xef), Uint8Array.of(0x12, 0x34)];
    expect(hex(computeParity(chunks, 1)[0])).toBe(hex(computeParity(chunks, 2)[0]));
  });

  it('emits min(k, n) symbols, so n=1 gets exactly one', () => {
    const chunks = [Uint8Array.of(7, 7, 7)];
    const p = computeParity(chunks, 2);
    expect(p).toHaveLength(1);
    expect(hex(p[0])).toBe(hex(chunks[0]));
  });

  it('returns nothing for k=0 and rejects out-of-range k', () => {
    const chunks = [Uint8Array.of(1), Uint8Array.of(2)];
    expect(computeParity(chunks, 0)).toEqual([]);
    expect(() => computeParity(chunks, MAX_PARITY_SYMBOLS + 1)).toThrow(WireError);
    expect(() => computeParity(chunks, -1)).toThrow(WireError);
  });

  it('refuses frames with more chunks than the code covers', () => {
    const chunks = Array.from({ length: MAX_PARITY_DATA_CHUNKS + 1 }, () => Uint8Array.of(1));
    expect(() => computeParity(chunks, 2)).toThrow(WireError);
  });
});

describe('recoverChunks', () => {
  // The load-bearing correctness proof: every supported n, every erasure
  // pair among the n+k transmitted chunks.
  // Compute-bound rather than IO-bound, so it gets explicit headroom: a
  // 2-core shared runner is several times slower than a dev machine, and the
  // default 5 s says "hung" about a test that is simply working.
  it('recovers every erasure pair it has capacity for', { timeout: 30_000 }, () => {
    let seed = 1;
    const rnd = () => {
      seed = (seed * 1103515245 + 12345) & 0x7fffffff;
      return seed & 0xff;
    };
    for (const n of [1, 2, 3, 9, 17, 64, 255]) {
      const width = 32;
      const orig = Array.from({ length: n }, (_, i) =>
        Uint8Array.from({ length: i === n - 1 ? width / 2 + 1 : width }, rnd),
      );
      const frameBytes = (n - 1) * width + orig[n - 1].length;
      const k = n < 2 ? 1 : 2;
      const parity = computeParity(orig, k);
      const total = n + parity.length;
      // Exhaustive only where it is cheap. The pair count is quadratic and
      // the work per pair linear, so the large n are seconds of JS for no
      // extra coverage — and they were 5 s on a 2-core CI runner, which is
      // what this threshold is set by.
      //
      // The exhaustive proof lives in the Go mirror, which runs every pair for
      // every n up to 255 in ~1.4 s. This mirror's job is to show the TS codec
      // AGREES, so above the threshold it samples the boundary pairs (first,
      // last, either side of the data/parity split, wrap-adjacent) instead —
      // and the coefficient-distinctness property those boundaries guard is
      // proven exhaustively by the gfPow2 test above regardless.
      const interesting = (i: number) =>
        n <= 17 || i < 2 || i >= total - 3 || i === n - 1 || i === n || i % 37 === 0;
      for (let a = 0; a < total; a++) {
        if (!interesting(a)) continue;
        for (let b = a; b < total; b++) {
          if (!interesting(b)) continue;
          const chunks: (Uint8Array | null)[] = [...orig];
          const par: (Uint8Array | null)[] = [...parity];
          for (const pos of [a, b]) {
            if (pos < n) chunks[pos] = null;
            else par[pos - n] = null;
          }
          const lost = chunks.filter((c) => c === null).length;
          const surviving = par.filter((p) => p !== null).length;
          if (lost > surviving) {
            expect(() => recoverChunks(chunks, par, frameBytes), `n=${n} (${a},${b})`).toThrow(WireError);
            continue;
          }
          recoverChunks(chunks, par, frameBytes);
          for (let i = 0; i < n; i++) {
            expect(hex(chunks[i]!), `n=${n} erasures (${a},${b}) chunk ${i}`).toBe(hex(orig[i]));
          }
        }
      }
    }
  });

  it('trims padding off a recovered short final chunk', () => {
    const orig = [Uint8Array.of(1, 2, 3, 4), Uint8Array.of(5, 6, 7, 8), Uint8Array.of(9)];
    const parity = computeParity(orig, 2);
    const chunks: (Uint8Array | null)[] = [orig[0], orig[1], null];
    recoverChunks(chunks, parity, 9);
    expect(Array.from(chunks[2]!)).toEqual([9]);
  });

  it('is a no-op on a complete frame and throws without parity', () => {
    const orig = [Uint8Array.of(1, 2), Uint8Array.of(3, 4)];
    expect(() => recoverChunks([...orig], [], 4)).not.toThrow();
    expect(() => recoverChunks([null, orig[1]], [], 4)).toThrow(WireError);
  });

  it('rejects a frameBytes that cannot describe the block', () => {
    const orig = [Uint8Array.of(1, 2, 3, 4), Uint8Array.of(5, 6)];
    const parity = computeParity(orig, 2);
    expect(() => recoverChunks([null, orig[1]], parity, 9999)).toThrow(WireError);
  });
});

describe('ParityChunk wire format', () => {
  it('round-trips', () => {
    const header = { frameId: 0x01020304, parityIndex: 1, chunkCount: 9, frameBytes: 8640 };
    const payload = new Uint8Array(64).fill(0xa5);
    const dgram = encodeParityChunk(header, payload);
    expect(dgram.length).toBe(PARITY_CHUNK_HEADER_SIZE + payload.length);
    const got = parseParityChunk(dgram);
    expect(got.header).toEqual(header);
    expect(hex(got.payload)).toBe(hex(payload));
  });

  it('matches the Go golden vector', () => {
    const dgram = encodeParityChunk(
      { frameId: 0x01020304, parityIndex: 1, chunkCount: 9, frameBytes: 8640 },
      Uint8Array.of(0xde, 0xad, 0xbe, 0xef),
    );
    expect(hex(dgram)).toBe('010e01020304010009000021c0deadbeef');
  });

  it('fits MAX_DATAGRAM_SIZE at a full payload — the common case, not an edge case', () => {
    const payload = new Uint8Array(MAX_CHUNK_PAYLOAD).fill(0x5a);
    const dgram = encodeParityChunk({ frameId: 1, parityIndex: 0, chunkCount: 9, frameBytes: 9 * MAX_CHUNK_PAYLOAD }, payload);
    expect(dgram.length).toBe(PARITY_CHUNK_HEADER_SIZE + MAX_CHUNK_PAYLOAD);
    expect(dgram.length).toBeLessThanOrEqual(MAX_DATAGRAM_SIZE);
    expect(parseParityChunk(dgram).payload.length).toBe(MAX_CHUNK_PAYLOAD);
  });

  it('rejects malformed datagrams', () => {
    const good = encodeParityChunk({ frameId: 1, parityIndex: 0, chunkCount: 4, frameBytes: 16 }, Uint8Array.of(1, 2, 3, 4));
    const mutate = (f: (b: Uint8Array) => Uint8Array) => f(Uint8Array.from(good));
    expect(() => parseParityChunk(good.subarray(0, PARITY_CHUNK_HEADER_SIZE - 1))).toThrow(WireError);
    expect(() => parseParityChunk(mutate((b) => { b[0] = 0x02; return b; }))).toThrow(WireError);
    expect(() => parseParityChunk(mutate((b) => { b[1] = TYPE_VIDEO_CHUNK; return b; }))).toThrow(WireError);
    expect(() => parseParityChunk(mutate((b) => { b[7] = 0; b[8] = 0; return b; }))).toThrow(WireError);
    expect(() => parseParityChunk(mutate((b) => { b[7] = 0xff; b[8] = 0xff; return b; }))).toThrow(WireError);
    expect(() => parseParityChunk(mutate((b) => { b[6] = MAX_PARITY_SYMBOLS; return b; }))).toThrow(WireError);
    const oversize = new Uint8Array(MAX_DATAGRAM_SIZE + 1);
    oversize.set(good);
    expect(() => parseParityChunk(oversize)).toThrow(WireError);
  });

  it('rejects an oversize payload at encode', () => {
    expect(() =>
      encodeParityChunk({ frameId: 1, parityIndex: 0, chunkCount: 2, frameBytes: 4 }, new Uint8Array(MAX_CHUNK_PAYLOAD + 1)),
    ).toThrow(WireError);
  });
});

describe('RelayCapabilities wire format', () => {
  it('round-trips and matches the Go golden vector', () => {
    const caps = { flags: CAP_PARITY_CHUNKS, parityLevel: 2 };
    const buf = encodeRelayCapabilities(caps);
    expect(buf.length).toBe(RELAY_CAPABILITIES_SIZE);
    expect(hex(buf)).toBe('010f000102');
    expect(parseRelayCapabilities(buf)).toEqual(caps);
  });

  it('rejects malformed messages', () => {
    const good = encodeRelayCapabilities({ flags: CAP_PARITY_CHUNKS, parityLevel: 1 });
    expect(() => parseRelayCapabilities(good.subarray(0, RELAY_CAPABILITIES_SIZE - 1))).toThrow(WireError);
    expect(() => parseRelayCapabilities(new Uint8Array([...good, 0]))).toThrow(WireError);
    expect(() => parseRelayCapabilities(Uint8Array.of(0x02, 0x0f, 0, 1, 1))).toThrow(WireError);
    expect(() => parseRelayCapabilities(Uint8Array.of(WIRE_VERSION, TYPE_VIDEO_CHUNK, 0, 1, 1))).toThrow(WireError);
    expect(() => parseRelayCapabilities(Uint8Array.of(WIRE_VERSION, 0x0f, 0, 1, MAX_PARITY_SYMBOLS + 1))).toThrow(WireError);
  });
});

describe('parity parsing never panics', () => {
  it('survives random bytes', () => {
    let seed = 7;
    const rnd = () => {
      seed = (seed * 1103515245 + 12345) & 0x7fffffff;
      return seed & 0xff;
    };
    for (let i = 0; i < 5000; i++) {
      const buf = Uint8Array.from({ length: i % 64 }, rnd);
      try {
        parseParityChunk(buf);
      } catch (e) {
        expect(e).toBeInstanceOf(WireError);
      }
      try {
        parseRelayCapabilities(buf);
      } catch (e) {
        expect(e).toBeInstanceOf(WireError);
      }
    }
  });
});
